### PR TITLE
ForceAtlas2 like in Gephi

### PR DESCRIPTION
**Pull request by SemyonSinchenko**
Hi!
I've implemented a ForceAtlas2 algorithm. I really think that it is one of the most popular and modern force-directed layout algorithm. I tried to save all features from the original paper but also I tried to follow the style of this great library.

Original paper:
https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0098679

### CODE REVIEW

File: jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+            }
+        }
+
+        if (Double.isNaN(force1) || Double.isNaN(force2))
+            throw new IllegalArgumentException(
+                    "Unexpected mathematical result in FRLayout:calcPositions");
+
+        if (!v1_locked) {
+            Point fvd1 = getFRData(vertex1);
+            frVertexData.put(vertex1, fvd1.add(-force1 * xDelta, -force1 * xDelta));
```
Comment (tomnelson) [inline jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java]: I think this should be:
` frVertexData.put(vertex1, fvd1.add(-force1 * xDelta, -force1 * yDelta));`
(yDelta in the 2nd argument)

File: jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+        }
+        if (!v2_locked) {
+            Point fvd2 = getFRData(vertex2);
+            frVertexData.put(vertex2, fvd2.add(force2 * xDelta, force2 * xDelta));
```
Comment (tomnelson) [inline jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java]: same as above comment

File: jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java (hunk 1 above)
Comment (SemyonSinchenko) [reply jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java]: I'm sorry about it. I've tested my local version and made these mistakes when move it to repo.

**Comment by SemyonSinchenko** [comment]
Yes I was testing it. Not exactly this code but my own local version in my project. I wrote it for myself (because the FR-algorithm was giving me really bad results: all nodes was in one big heap) but I think it may be useful here too. Of course you can move it in any other PR! I see it's all right here!

**Comment by tomnelson** [comment]
I'd like to merge your PR with the changes that I suggested and showed in the PR07 branch. Would you like to update your branch for this PR so I can merge it and include ForceAtlas2 in the sample programs?

**Comment by SemyonSinchenko** [comment]
I'm sorry! I didn't understand You right. I've finished fixes but I see that You already merged it.
One more thing: it seems that there are two another errors.

- https://github.com/tomnelson/jungrapht-visualization/blob/42584b82232c1b85b48b6bd51dff6bbb6e85fa46/jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java#L388
- https://github.com/tomnelson/jungrapht-visualization/blob/42584b82232c1b85b48b6bd51dff6bbb6e85fa46/jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java#L392

It seems that I forgot divide xDelta and yDelta by dist. They aren't really Cos and Sin now but they must be! Really soory about it!

**Comment by tomnelson** [comment]
I was that your branch was gone from your fork and that you had merged into your master so I thought it would be too much trouble for you to make the changes. 
Do you want to submit another PR with the corrections to divide by dist? (I can just make the changes if you prefer). 
Thanks again for contributing this and for all your work.

**Comment by SemyonSinchenko** [comment]
I'll make one another PR. Give a 30 minutes :)

**Pull request merged**